code/bmpman/bmpman.cpp

```diff
@@ -2076,6 +2109,61 @@ void bm_page_in_start() {
 	gr_bm_page_in_start();
 }
 
+/**
+ * Stuffs Pathtypes with an additional path leading to a subdir of the same name as the given file (minus file extension).
+ * Used for allowing frames of .eff animations to be found in their own subdirectories.
+ *
+ * Example: if called with "debris.eff" and that file is in data/maps, then Pathtypes[CF_TYPE_TEMP_SUBDIR_LOOKUP].path
+ * will be stuffed with data/maps/debris, allowing data/maps/debris/debris_0000.dds (etc) to be found
+ */
+bool set_temp_subdir_pathtype(const char *filename) {
+	// Gets the absolute path of the file, removes the part that is identical with the
+	// absolute root path and appends the filename to it (without extension), giving us
+	// a relative path that can be temporarily stuffed into Pathtypes so the frames can be found
+
+	char fullpath[MAX_PATH]; // Absolute path to the given file
+	char fullrootpath[MAX_PATH]; // Absolute root path
+	int size, offset; // Needed for cf_find_file_location, but not used
+
+	cf_create_default_path_string(fullrootpath, sizeof(fullrootpath) - 1, CF_TYPE_ROOT, NULL);
+	if (!cf_find_file_location(filename, CF_TYPE_ANY, sizeof(fullpath)-1, fullpath, &size, &offset)) {
+		mprintf(("BMPMAN: Failed to set temporary pathtype for %s: file not found!\n", filename));
+		return false;
+	}
+
+#ifndef NDEBUG
+	if (strlen(fullpath) < strlen(fullrootpath)) {
+		mprintf(("BMPMAN: Path to %s was shorter than the root path, programmer thought it could never happen!\n%s\n%s\n", filename, fullpath, fullrootpath));
+		return false;
+	}
+#endif
+
+	mprintf(("BMPMAN: Setting temporary pathtype for %s... ", filename));
+
+	char *trimmed_path; // The differing portion of fullpath and fullrootpath
+	trimmed_path = &fullpath[strlen(fullrootpath)];
+
+	char *extpos = strstr(trimmed_path, "."); // Where the file extension begins
+	char basepath[MAX_PATH]; // This will hold the final, relative path
+
+	// Copy the relative path of the given file to basepath, excluding the extension
+	strncpy(basepath, trimmed_path, extpos-trimmed_path);
+
+	// Make sure it's null-terminated right
+	basepath[(int)(extpos-trimmed_path)] = '\0';
+
+	// Free the old string first, if it's not the const initial empty string
+	if (Pathtypes[CF_TYPE_TEMP_SUBDIR_LOOKUP].path[0] != '\0') {
+		vm_free(Pathtypes[CF_TYPE_TEMP_SUBDIR_LOOKUP].path);
+	}
+
+	Pathtypes[CF_TYPE_TEMP_SUBDIR_LOOKUP].path = vm_strdup(basepath);
```

Comment:
The memory from this `vm_strdup` call is leaked at the end of the program. Previous values are freed but the last value that is allocated is not freed anywhere.
It's also a bit weird to mix string literals and heap-allocated string. At first `path` is assigned the empty string literal but here you assign a heap allocated string to `path`.